src/test/spec/v2_runner/mod.rs

```diff
@@ -232,220 +226,282 @@ async fn run_v2_test(path: std::path::PathBuf, test_file: TestFile) {
             }
         }
 
+        // Persist fail point guards so they disable post-test.
         let mut fail_point_guards: Vec<FailPointGuard> = Vec::new();
-        if let Some(fail_point) = test.fail_point {
+        if let Some(fail_point) = &test.fail_point {
             fail_point_guards.push(fail_point.enable(client.deref(), None).await.unwrap());
         }
 
+        // Start the test sessions
         let options = match test.session_options {
             Some(ref options) => options.get("session0").cloned(),
             None => None,
         };
-        let mut session0 = Some(client.start_session(options).await.unwrap());
-        let session0_lsid = session0.as_ref().unwrap().id().clone();
+        let session0 = Some(client.start_session(options).await.unwrap());
 
         let options = match test.session_options {
             Some(ref options) => options.get("session1").cloned(),
             None => None,
         };
-        let mut session1 = Some(client.start_session(options).await.unwrap());
-        let session1_lsid = session1.as_ref().unwrap().id().clone();
+        let session1 = Some(client.start_session(options).await.unwrap());
 
-        for operation in test.operations {
-            let db = match &operation.database_options {
-                Some(options) => client.database_with_options(&db_name, options.clone()),
-                None => client.database(&db_name),
-            };
-            let coll = match &operation.collection_options {
-                Some(options) => db.collection_with_options(&coll_name, options.clone()),
-                None => db.collection(&coll_name),
-            };
+        Self {
+            description: test.description.clone(),
+            ns: Namespace {
+                db: db_name,
+                coll: coll_name,
+            },
+            internal_client: internal_client.clone(),
+            client,
+            fail_point_guards,
+            session0,
+            session1,
+        }
+    }
 
-            let session = match operation.session.as_deref() {
-                Some("session0") => session0.as_mut(),
-                Some("session1") => session1.as_mut(),
-                Some(other) => panic!("unknown session name: {}", other),
-                None => None,
+    async fn run_operation(
+        &mut self,
+        operation: &Operation,
+    ) -> Option<Result<Option<bson::Bson>, crate::error::Error>> {
```

Comment:
The return type here is a little complex because along with succeeding or failing an operation can also be skipped.